gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/UserQuotaManager.java

```diff
@@ -0,0 +1,197 @@
+      throw new IOException(requesterMessage.toString());
+    }
+  }
+
+  /**
+   * Increment quota by one for the given map and key.
+   * @return a negative number if quota is already reached for this user
+   *         a positive number if the quota is not reached for this user
+   *         the absolute value of the number is the used quota before this increment request
+   *         0 if quota usage is not changed
+   */
+  private int incrementJobCountAndCheckUserQuota(Map<String, Integer> quotaMap, String user, Dag.DagNode<JobExecutionPlan> dagNode) {
+    String key = DagManagerUtils.getUserQuotaKey(user, dagNode);
+
+    // Only increment job count for first attempt, since job is considered running between retries
+    if (dagNode.getValue().getCurrentAttempts() != 1) {
+      return 0;
+    }
+
+    Integer currentCount;
+    do {
+      currentCount = quotaMap.get(key);
+    } while (currentCount == null ? quotaMap.putIfAbsent(key, 1) != null : !quotaMap.replace(key, currentCount, currentCount + 1));
+
+    if (currentCount == null) {
+      currentCount = 0;
+    }
+
+    if (currentCount >= getQuotaForUser(user)) {
+      return -currentCount; // increment must have crossed the quota
+    } else {
+      return currentCount;
+    }
+  }
+
+  /**
+   * Decrement the quota by one for the proxy user and requesters corresponding to the provided {@link Dag.DagNode}.
+   * Returns true if the dag existed in the set of running dags and was removed successfully
+   */
+  public boolean releaseQuota(Dag.DagNode<JobExecutionPlan> dagNode) {
+    Boolean val = runningDagIds.remove(DagManagerUtils.generateDagId(dagNode));
+    if (val == null) {
+      return false;
+    }
+    String proxyUser = ConfigUtils.getString(dagNode.getValue().getJobSpec().getConfig(), AzkabanProjectConfig.USER_TO_PROXY, null);
+    if (proxyUser != null) {
+      String proxyUserKey = DagManagerUtils.getUserQuotaKey(proxyUser, dagNode);
+      decrementQuotaUsage(proxyUserToJobCount, proxyUserKey);
+    }
+    String serializedRequesters = DagManagerUtils.getSerializedRequesterList(dagNode);
+    if (serializedRequesters != null) {
+      try {
+        for (ServiceRequester requester : RequesterService.deserialize(serializedRequesters)) {
+          String requesterKey = DagManagerUtils.getUserQuotaKey(requester.getName(), dagNode);
+          decrementQuotaUsage(requesterToJobCount, requesterKey);
+        }
+      } catch (IOException e) {
+        log.error("Failed to release quota for requester list " + serializedRequesters, e);
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private void decrementQuotaUsage(Map<String, Integer> quotaMap, String user) {
+    Integer currentCount;
+    if (user == null) {
+      return;
+    }
+    do {
+      currentCount = quotaMap.get(user);
+    } while (currentCount != null && currentCount > 0 && !quotaMap.replace(user, currentCount, currentCount - 1));
```

Comment:
maybe a comment, like, "mods must be thread-safe since DAGs on different DMThreads may concern the same user"?